### PR TITLE
alsa: Option to ignore all ALSA plugin devices.

### PR DESCRIPTION
**Pull request by dmitrykos**
Option to ignore all ALSA plugin devices by using PA_ALSA_IGNORE_ALL_PLUGINS environment variable. 

Fixed logic of guarding assertion in BuildDeviceList() when only one PCM device is available (may occur if PA_ALSA_IGNORE_ALL_PLUGINS is defined).

Ignoring ALSA plugin devices is useful when using using PortAudio with ALSA interface in embedded systems like [Yocto Linux](http://www.yoctoproject.org). Some plugins are causing segfaults and practically may not be needed at all for the developer. By defining PA_ALSA_IGNORE_ALL_PLUGINS developer is able to switch off Alsa Plugin devices, only PCM devices will exist in this case (internal sound card, USB DAC, ...).

### CODE REVIEW

File: src/hostapi/alsa/pa_linux_alsa.c [hunk 1]
```diff
@@ -1063,6 +1063,7 @@ static PaError PaAlsa_StrDup( PaAlsaHostApiRepresentation *alsaApi,
  */
 static int IgnorePlugin( const char *pluginId )
 {
+#ifndef PA_ALSA_IGNORE_ALL_PLUGINS
```
Comment (RossBencina) [inline src/hostapi/alsa/pa_linux_alsa.c:1066]: Is there a reason why you did the #ifndef here rather than at the call site? :

https://github.com/PortAudio/portaudio/blob/e1f0c230aac0d9aa0d45b4f8600061de6f14c62f/src/hostapi/alsa/pa_linux_alsa.c#L1388

Maybe the debug logging should not be suppressed.

File: src/hostapi/alsa/pa_linux_alsa.c [hunk 2]
```diff
@@ -1450,7 +1456,7 @@ static PaError BuildDeviceList( PaAlsaHostApiRepresentation *alsaApi )
 
         PA_ENSURE( FillInDevInfo( alsaApi, hwInfo, blocking, devInfo, &devIdx ) );
     }
-    assert( devIdx < numDeviceNames );
+    assert( devIdx <= numDeviceNames );
```
Comment (RossBencina) [inline src/hostapi/alsa/pa_linux_alsa.c:1459]: What is the reason for this change? It is not mentioned in the commit message.

File: src/hostapi/alsa/pa_linux_alsa.c (hunk 1 above)
Comment (dmitrykos) [reply src/hostapi/alsa/pa_linux_alsa.c:1066]: I did not want to pollute the code using an existing `IgnorePlugin` function, in fact modifying `IgnorePlugin` is perfectly fine for the logic of `ifdef` as it already ignoring some plugins and with `#ifndef PA_ALSA_IGNORE_ALL_PLUGINS` will ignore all plugins, that is the goal of the proposed change.

File: src/hostapi/alsa/pa_linux_alsa.c (hunk 2 above)
Comment (dmitrykos) [reply src/hostapi/alsa/pa_linux_alsa.c:1459]: It is mentioned `Fixed logic of guarding assertion in BuildDeviceList() when only one PCM device is available (may occur if PA_ALSA_IGNORE_ALL_PLUGINS is defined).`. 

When `PA_ALSA_IGNORE_ALL_PLUGINS` is defined, e.g. user compiled PA to ignore all ALSA plugins, then `assert( devIdx < numDeviceNames )` will become true and process will stop.

Comment (dmitrykos) [reply src/hostapi/alsa/pa_linux_alsa.c:1459]: The algorithm of the failing assert is the following:

Assume device has only USB DAC attached as playback device.

1. `numDeviceNames` becomes 1
2. `FillInDevInfo` (line 1457) increments `devIdx` to 1
3. loop at line 1448 `for( i = 0, devIdx = 0; i < numDeviceNames; ++i )` exits as there is only one PCM device (`numDeviceNames == 1`)
4. assert condition becomes false `assert(1 < 1)`

**Comment by dmitrykos** [comment]
The discussed assert (https://github.com/PortAudio/portaudio/blob/master/src/hostapi/alsa/pa_linux_alsa.c#L1454) is also affecting the issue #294 which can be closed by this patch.

Close #294

File: CMakeLists.txt [hunk 3]
```diff
@@ -297,12 +297,16 @@ ELSE()
       OPTION(PA_USE_ALSA "Enable support for ALSA" OFF)
     ENDIF()
     IF(PA_USE_ALSA)
+      OPTION(PA_ALSA_IGNORE_ALL_PLUGINS "Ignore all ALSA plugin devices" OFF)
```
Comment (philburk) [inline CMakeLists.txt:300]: I notice that several ALSA options are controlled by environment variables. See multiple calls to getenv() in "pa_linux_alsa.c". That would allow a user to disable plugin devices for a precompiled app. If the plugin devices are specific to a machine then using getenv() might be better than disabling plugins for the app on all machines.

File: src/hostapi/alsa/pa_linux_alsa.c (hunk 1 above)
Comment (philburk) [reply src/hostapi/alsa/pa_linux_alsa.c:1066]: If we used a getenv() then we could just skip the entire block that starts at 1358.

Comment (philburk) [reply src/hostapi/alsa/pa_linux_alsa.c:1066]: If we used a getenv() then we could just skip the entire block that starts at 1363.
Maybe something like:

    if (getenv("PA_ALSA_IGNORE_ALL_PLUGINS")
            && atoi( getenv( "PA_ALSA_IGNORE_ALL_PLUGINS" ) ) )
    {
                PA_DEBUG(( "%s: Ignoring all ALSA plugin devices\n", __FUNCTION__ ));
    }
    else if( ( res = alsa_snd_config_search( *alsa_snd_config, "pcm", &topNode ) ) >= 0 )

Comment (philburk) [reply src/hostapi/alsa/pa_linux_alsa.c:1066]: @dmitrykos - What do you think about using getenv()?
Using getenv() makes it machine specific so users can configure it. Plugins are also machine specific so it seems more appropriate to use getenv().

Comment (dmitrykos) [reply src/hostapi/alsa/pa_linux_alsa.c:1066]: @philburk , thank you for your comment. I agree, will update implementation accordingly.

**Comment by dmitrykos** [comment]
Changed implementation to the usage of environment variable instead of compile-time define as per our discussion.

**Review by philburk — APPROVED** [review]
Thanks for using getenv().

**Review by RossBencina — APPROVED** [review]
